google/cloud/opentelemetry/resource_detector.cc

```diff
@@ -0,0 +1,241 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/opentelemetry/resource_detector.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/compute_engine_util.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/internal/rest_client.h"
+#include "google/cloud/log.h"
+#include <nlohmann/json.hpp>
+#include <opentelemetry/sdk/resource/resource.h>
+#include <opentelemetry/sdk/resource/semantic_conventions.h>
+
+namespace google {
+namespace cloud {
+namespace otel {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+namespace sc = opentelemetry::sdk::resource::SemanticConventions;
+
+// The metadata server returns fully qualified names. (e.g. a zone may be
+// "projects/p/zones/us-central1-a"). Return the IDs only.
+std::string Tail(std::string const& value) {
+  auto const pos = value.rfind('/');
+  return value.substr(pos + 1);
+}
+
+std::string MD(nlohmann::json const& json,
+               std::vector<std::string> const& keys) {
+  auto const* j = &json;
+  for (auto const& k : keys) {
+    if (!j->contains(k)) return {};
```

Comment:
I think here you want something like:

```cc
auto const f = j->find(k);
if (f == j->end()) return {};
j = &(*f);
```

But even that is a bit weird and obscures what you are doing.  Maybe:

```cc
std::string FindRecursive(nlohmann::json const& json, std::deque<std::string> keys) {
  if (!keys.empty()) {
    auto const l = json.find(keys.front());
    if (l ==json.end()) return {};
    keys.pop_front();
    return FindRecursive(*l, std::move(keys));
  }
  if (json.is_string()) return json.get<std::string>();
  if (json.is_number()) return std::to_string(json.get<std::int64_t>());
  return {};
}
```